File: AgentforceSDK-ReactNative-Bridge/android/src/main/java/com/salesforce/android/reactagentforce/AppearanceConfiguration.kt
package com.salesforce.android.reactagentforce

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Rect
import android.graphics.drawable.Drawable
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.graphics.painter.BitmapPainter
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReadableType
import com.salesforce.android.agentforcesdk.components.theme.AgentforceColorToken
import com.salesforce.android.agentforcesdk.components.theme.AgentforceDisplayNameType
import com.salesforce.android.agentforcesdk.components.theme.AgentforceIcon
import com.salesforce.android.agentforcesdk.components.theme.AgentforceIconType
import com.salesforce.android.agentforcesdk.components.theme.AgentforceThemeMode
import com.salesforce.android.agentforcesdk.components.theme.AgentforceThemeOverrides
import com.salesforce.android.agentforcesdk.components.theme.AgentforceTheming
import com.salesforce.android.agentforcesdk.components.theme.AgentforceTypography
import androidx.compose.ui.graphics.Color

internal object AppearanceConfiguration {
    fun theming(config: ReadableMap, context: Context): AgentforceTheming? {
        if (!config.hasKey("appearance") || config.getType("appearance") == ReadableType.Null) return null
        val raw = config.getMap("appearance") ?: throw IllegalArgumentException("appearance must be an object")
        val overrides = AgentforceThemeOverrides(
            lightColors = colors(raw.getMapOrNull("lightColors"), "lightColors"),
            darkColors = colors(raw.getMapOrNull("darkColors"), "darkColors"),
            displayNames = displayNames(raw.getMapOrNull("displayNames")),
            icons = icons(raw.getMapOrNull("icons"), context),
            themeMode = themeMode(raw.getStringOrNull("themeMode")),
            typography = typography(raw.getMapOrNull("typography"), context)
        )
        return AgentforceTheming.Overrides(overrides)
    }

    private fun colors(values: ReadableMap?, field: String): Map<AgentforceColorToken, Color> {
        if (values == null) return emptyMap()
        val result = mutableMapOf<AgentforceColorToken, Color>()
        val keys = values.keySetIterator()
        while (keys.hasNextKey()) {
            val key = keys.nextKey()
            val token = AgentforceColorToken.entries.firstOrNull { it.key == key }
                ?: throw IllegalArgumentException("Unsupported Android color token '$key'")
            val value = values.getString(key)
                ?: throw IllegalArgumentException("appearance.$field.$key must be a color string")
            result[token] = color(value, "appearance.$field.$key")
        }
        return result
    }

    private fun color(value: String, field: String): Color {
        val hex = value.trim()
        if (!Regex("^#[0-9a-fA-F]{6}([0-9a-fA-F]{2})?$").matches(hex)) {
            throw IllegalArgumentException("$field must be #RRGGBB or #AARRGGBB")
        }
        val parsed = hex.drop(1).toLong(16)
        return if (hex.length == 7) Color(0xFF000000 or parsed) else Color(parsed)
    }

    private fun displayNames(values: ReadableMap?): Map<AgentforceDisplayNameType, String> {
        if (values == null) return emptyMap()
        val result = mutableMapOf<AgentforceDisplayNameType, String>()
        val keys = values.keySetIterator()
        while (keys.hasNextKey()) {
            val key = keys.nextKey()
            val type = AgentforceDisplayNameType.entries.firstOrNull { it.name.replaceFirstChar(Char::lowercase) == key }
                ?: throw IllegalArgumentException("Unsupported Android display-name token '$key'")
            val value = values.getString(key)
            if (value.isNullOrEmpty()) throw IllegalArgumentException("appearance.displayNames.$key must not be empty")
            result[type] = value
        }
        return result
    }

    private fun icons(values: ReadableMap?, context: Context): Map<AgentforceIconType, AgentforceIcon> {
        if (values == null) return emptyMap()
        val result = mutableMapOf<AgentforceIconType, AgentforceIcon>()
        val keys = values.keySetIterator()
        while (keys.hasNextKey()) {
            val key = keys.nextKey()
            val type = AgentforceIconType.entries.firstOrNull { it.name.replaceFirstChar(Char::lowercase) == key }
                ?: throw IllegalArgumentException("Unsupported Android icon token '$key'")
            val platform = values.getMap(key)?.getMap("android")
                ?: throw IllegalArgumentException("appearance.icons.$key requires android.light")
            val light = platform.getString("light")
                ?: throw IllegalArgumentException("appearance.icons.$key requires android.light")
            val lightPainter = painter(light, context, "appearance.icons.$key.android.light")
            val darkPainter = platform.getString("dark")?.let {
                painter(it, context, "appearance.icons.$key.android.dark")
            }
            result[type] = AgentforceIcon(lightPainter, darkPainter ?: lightPainter)
        }
        return result
    }

    private fun painter(resourceName: String, context: Context, field: String): BitmapPainter {
        val id = context.resources.getIdentifier(resourceName, "drawable", context.packageName)
        if (id == 0) throw IllegalArgumentException("Android drawable '$resourceName' for $field was not found")
        val bitmap = BitmapFactory.decodeResource(context.resources, id)
            ?: context.getDrawable(id)?.let(::drawableBitmap)
            ?: throw IllegalArgumentException("Android drawable '$resourceName' for $field could not be decoded")
        return BitmapPainter(bitmap.asImageBitmap())
    }

    /** Vector drawables cannot be decoded by BitmapFactory, so rasterize them for Compose. */
    internal fun drawableBitmap(drawable: Drawable): Bitmap {
        val previousBounds = Rect(drawable.bounds)
        val width = drawable.intrinsicWidth.takeIf { it > 0 } ?: 1
        val height = drawable.intrinsicHeight.takeIf { it > 0 } ?: 1
        return Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888).also { bitmap ->
            try {
                drawable.setBounds(0, 0, width, height)
                drawable.draw(Canvas(bitmap))
            } finally {
                drawable.bounds = previousBounds
            }
        }
    }

    private fun themeMode(value: String?): AgentforceThemeMode? = when (value) {
        null -> null
        "system" -> AgentforceThemeMode.SYSTEM
        "light" -> AgentforceThemeMode.LIGHT
        "dark" -> AgentforceThemeMode.DARK
        else -> throw IllegalArgumentException("appearance.themeMode must be 'system', 'light', or 'dark'")
    }

    private fun typography(raw: ReadableMap?, context: Context): com.salesforce.android.sharedui.theme.Typography? {
        if (raw == null) return null
        val base = raw.getMapOrNull("fontFamily")?.let { family(it, context) }?.let {
            AgentforceTypography.Default.copyWithFontFamily(it)
        } ?: AgentforceTypography.Default
        val styles = raw.getMapOrNull("styles") ?: return base
        val overrides = mutableMapOf<String, TextStyle>()
        val keys = styles.keySetIterator()
        while (keys.hasNextKey()) {
            val name = keys.nextKey()
            val current = base[name] ?: throw IllegalArgumentException("Unsupported Android typography style '$name'")
            val rawStyle = styles.getMap(name) ?: throw IllegalArgumentException("appearance.typography.styles.$name must be an object")
            val size = rawStyle.getDoubleOrNull("size")?.also {
                if (!it.isFinite() || it <= 0 || it > 512) throw IllegalArgumentException("appearance.typography.styles.$name.size must be between 0 and 512")
            }?.toFloat()?.sp ?: current.fontSize
            val weight = rawStyle.getIntOrNull("weight")?.let { fontWeight(it) } ?: current.fontWeight
            val styleFamily = rawStyle.getMapOrNull("fontFamily")?.let { family(it, context) } ?: current.fontFamily
            overrides[name] = current.copy(fontSize = size, fontWeight = weight, fontFamily = styleFamily)
        }
        return base.mergeWith(overrides)
    }

    private fun family(raw: ReadableMap, context: Context): FontFamily {
        return when (raw.getString("type")) {
            "generic" -> when (raw.getString("family")) {
                "default" -> FontFamily.Default
                "sans-serif" -> FontFamily.SansSerif
                "serif" -> FontFamily.Serif
                "monospace" -> FontFamily.Monospace
                "cursive" -> FontFamily.Cursive
                else -> throw IllegalArgumentException("Unsupported generic font family")
            }
            "bundled" -> {
                val resources = raw.getMap("android")?.getMap("resources")
                    ?: throw IllegalArgumentException("Bundled Android font requires android.resources")
                val fonts = mutableListOf<Font>()
                val keys = resources.keySetIterator()
                while (keys.hasNextKey()) {
                    val weightKey = keys.nextKey()
                    val weight = fontWeight(weightKey.toIntOrNull() ?: throw IllegalArgumentException("Android font resource weight must be numeric"))
                    val name = resources.getString(weightKey)
                        ?: throw IllegalArgumentException("Android font resource name is required")
                    val id = context.resources.getIdentifier(name, "font", context.packageName)
                    if (id == 0) throw IllegalArgumentException("Android font resource '$name' was not found")
                    fonts += Font(id, weight)
                }
                if (fonts.isEmpty()) throw IllegalArgumentException("Bundled Android font requires at least one resource")
                FontFamily(*fonts.toTypedArray())
            }
            else -> throw IllegalArgumentException("appearance typography font family is invalid")
        }
    }

    private fun fontWeight(value: Int): FontWeight = when (value) {
        100, 200, 300, 400, 500, 600, 700, 800, 900 -> FontWeight(value)
        else -> throw IllegalArgumentException("Typography weight must be one of 100 through 900 in increments of 100")
    }

    private fun ReadableMap.getMapOrNull(key: String): ReadableMap? = if (hasKey(key) && getType(key) != ReadableType.Null) getMap(key) else null
    private fun ReadableMap.getStringOrNull(key: String): String? = if (hasKey(key) && getType(key) != ReadableType.Null) getString(key) else null
    private fun ReadableMap.getDoubleOrNull(key: String): Double? = if (hasKey(key) && getType(key) != ReadableType.Null) getDouble(key) else null
    private fun ReadableMap.getIntOrNull(key: String): Int? = getDoubleOrNull(key)?.toInt()
}

File: AgentforceSDK-ReactNative-Bridge/ios/Agentforce/AppearanceConfiguration.swift
import AgentforceSDK
import SwiftUI
import UIKit

enum AppearanceConfiguration {
    static func theming(from config: [String: Any]) throws -> AgentforceTheming? {
        guard let raw = config["appearance"] as? [String: Any] else { return nil }
        let light = try colors(raw["lightColors"], field: "lightColors")
        let dark = try colors(raw["darkColors"], field: "darkColors")
        let icons = try icons(raw["icons"])
        let displayNames = try displayNames(raw["displayNames"])
        let typography = try typography(raw["typography"])
        return .overrides(AgentforceThemeOverrides(
            light: light,
            dark: dark,
            icons: icons,
            displayNames: displayNames,
            fonts: typography.fonts,
            fontFamilyName: typography.family
        ))
    }

    static func themeMode(from config: [String: Any]) throws -> String? {
        guard let raw = config["appearance"] as? [String: Any], let mode = raw["themeMode"] as? String else {
            return nil
        }
        switch mode {
        case "system", "light", "dark": return mode
        default: throw AppearanceError.invalid("appearance.themeMode must be 'system', 'light', or 'dark'")
        }
    }

    static func hasOverrides(from config: [String: Any]) -> Bool {
        guard let raw = config["appearance"] as? [String: Any] else { return false }
        return raw["lightColors"] != nil || raw["darkColors"] != nil || raw["icons"] != nil ||
            raw["displayNames"] != nil || raw["typography"] != nil
    }

    static func themeManager(for mode: String) -> AgentforceThemeManager {
        switch mode {
        case "system": return AgentforceDefaultThemeManager(themeMode: .system)
        case "dark": return AgentforceDefaultThemeManager(themeMode: .dark)
        default: return AgentforceDefaultThemeManager(themeMode: .light)
        }
    }

    private static func colors(_ value: Any?, field: String) throws -> [AgentforceColorToken: Color] {
        guard let values = value as? [String: String] else { return [:] }
        return try values.reduce(into: [:]) { result, entry in
            guard let token = AgentforceColorToken(rawValue: entry.key) else {
                throw AppearanceError.invalid("Unsupported iOS color token '\(entry.key)'")
            }
            result[token] = try color(entry.value, field: "appearance.\(field).\(entry.key)")
        }
    }

    private static func color(_ value: String, field: String) throws -> Color {
        let hex = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard hex.hasPrefix("#"), hex.count == 7 || hex.count == 9,
              let number = UInt64(hex.dropFirst(), radix: 16) else {
            throw AppearanceError.invalid("\(field) must be #RRGGBB or #AARRGGBB")
        }
        let rgba: (UInt64, UInt64, UInt64, UInt64)
        if hex.count == 7 {
            rgba = ((number >> 16) & 0xff, (number >> 8) & 0xff, number & 0xff, 0xff)
        } else {
            rgba = ((number >> 16) & 0xff, (number >> 8) & 0xff, number & 0xff, (number >> 24) & 0xff)
        }
        return Color(.sRGB, red: Double(rgba.0) / 255, green: Double(rgba.1) / 255,
                     blue: Double(rgba.2) / 255, opacity: Double(rgba.3) / 255)
    }

    private static func icons(_ value: Any?) throws -> [AgentforceIconType: AgentforceIcon] {
        guard let values = value as? [String: Any] else { return [:] }
        return try values.reduce(into: [:]) { result, entry in
            guard let type = AgentforceIconType.allCases.first(where: { String(describing: $0) == entry.key }),
                  let platforms = entry.value as? [String: Any],
                  let ios = platforms["ios"] as? [String: String], let light = ios["light"] else {
                throw AppearanceError.invalid("appearance.icons.\(entry.key) requires ios.light")
            }
            guard let lightImage = UIImage(named: light) else {
                throw AppearanceError.invalid("iOS image asset '\(light)' for \(entry.key) was not found")
            }
            if let dark = ios["dark"] {
                guard let darkImage = UIImage(named: dark) else {
                    throw AppearanceError.invalid("iOS image asset '\(dark)' for \(entry.key) was not found")
                }
                result[type] = AgentforceIcon(lightModeIcon: Image(uiImage: lightImage), darkModeIcon: Image(uiImage: darkImage))
            } else {
                result[type] = AgentforceIcon(Image(uiImage: lightImage))
            }
        }
    }

    private static func displayNames(_ value: Any?) throws -> [AgentforceDisplayNameType: String] {
        guard let values = value as? [String: String] else { return [:] }
        return try values.reduce(into: [:]) { result, entry in
            guard let type = AgentforceDisplayNameType.allCases.first(where: { String(describing: $0) == entry.key }),
                  !entry.value.isEmpty else {
                throw AppearanceError.invalid("Unsupported or empty iOS display-name token '\(entry.key)'")
            }
            result[type] = entry.value
        }
    }

    private static func typography(_ value: Any?) throws -> (family: String?, fonts: [AgentforceFontStyle: AgentforceFontOverride]) {
        guard let raw = value as? [String: Any] else { return (nil, [:]) }
        let family = try fontFamily(raw["fontFamily"])
        guard let styles = raw["styles"] as? [String: [String: Any]] else { return (family, [:]) }
        let fonts = try styles.reduce(into: [AgentforceFontStyle: AgentforceFontOverride]()) { result, entry in
            guard let style = fontStyle(entry.key) else {
                throw AppearanceError.invalid("Unsupported iOS typography style '\(entry.key)'")
            }
            let size = entry.value["size"] as? NSNumber
            if let size, (!size.doubleValue.isFinite || size.doubleValue <= 0 || size.doubleValue > 512) {
                throw AppearanceError.invalid("appearance.typography.styles.\(entry.key).size must be between 0 and 512")
            }
            let weight = try fontWeight(entry.value["weight"])
            let styleFamily = try fontFamily(entry.value["fontFamily"])
            result[style] = AgentforceFontOverride(size: size.map { CGFloat($0.doubleValue) }, weight: weight, fontFamilyName: styleFamily)
        }
        return (family, fonts)
    }

    private static func fontStyle(_ value: String) -> AgentforceFontStyle? {
        // Android's public names include `Font` (for example,
        // `bodyFontScale1Regular`), while the iOS SDK omits it. Accept both
        // spellings so a shared JS appearance object maps to the same token.
        AgentforceFontStyle(rawValue: value) ??
            AgentforceFontStyle(rawValue: value.replacingOccurrences(of: "FontScale", with: "Scale"))
    }

    private static func fontFamily(_ value: Any?) throws -> String? {
        guard let raw = value as? [String: Any] else { return nil }
        switch raw["type"] as? String {
        case "generic":
            switch raw["family"] as? String {
            case "default", "sans-serif": return nil
            case "serif": return "Georgia"
            case "monospace": return "Menlo"
            case "cursive": return "Snell Roundhand"
            default: throw AppearanceError.invalid("Unsupported generic font family")
            }
        case "bundled":
            guard let ios = raw["ios"] as? [String: String], let name = ios["name"], UIFont(name: name, size: 12) != nil else {
                throw AppearanceError.invalid("Bundled iOS font must name a registered font")
            }
            return name
        case .none: return nil
        default: throw AppearanceError.invalid("appearance typography font family is invalid")
        }
    }

    private static func fontWeight(_ value: Any?) throws -> Font.Weight? {
        guard let value = value as? NSNumber else { return nil }
        switch value.intValue {
        case 100: return .ultraLight; case 200: return .thin; case 300: return .light
        case 400: return .regular; case 500: return .medium; case 600: return .semibold
        case 700: return .bold; case 800: return .heavy; case 900: return .black
        default: throw AppearanceError.invalid("Typography weight must be one of 100 through 900 in increments of 100")
        }
    }
}

enum AppearanceError: LocalizedError {
    case invalid(String)
    var errorDescription: String? { if case let .invalid(message) = self { return message }; return nil }
}
